Rolling-upgrade RecoveryIT tests are broken

The 6.x intake builds have been broken for most of today, with clusters failing to pass their wait conditions properly:

```
[ant:echo] ==> [Thu Nov 15 05:56:30 UTC 2018] checking health: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
     [echo] ==> [Thu Nov 15 05:56:30 UTC 2018] checking health: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
[ant:get] Getting: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
      [get] Getting: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
[ant:get] To: /var/lib/jenkins/workspace/elastic+elasticsearch+6.x+intake/qa/rolling-upgrade/build/cluster/v5.6.14#upgradedClusterTestCluster node0/cwd/wait.success
      [get] To: /var/lib/jenkins/workspace/elastic+elasticsearch+6.x+intake/qa/rolling-upgrade/build/cluster/v5.6.14#upgradedClusterTestCluster node0/cwd/wait.success

[ant:get] Error opening connection java.io.IOException: Server returned HTTP response code: 408 for URL: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
      [get] Error opening connection java.io.IOException: Server returned HTTP response code: 408 for URL: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
[ant:get] Error opening connection java.io.IOException: Server returned HTTP response code: 408 for URL: http://[::1]:35001/_cluster/health?wait_for_nodes=>=1&wait_for_status=yellow
```

This is from https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+intake/197/console although there are lots of other similar failures.

There's nothing obvious to me in the node logs themselves - the 3-node cluster looks to have formed fine. It does seem to reproduce locally, and @original-brownbear and I have been looking at this for a few hours without too much success. I will mute this test suite.

It is perhaps notable that all of the failures I can see have been in the two-thirds upgraded cluster or the fully-upgraded cluster. We've also seen cases where the cluster _does_ form but then times out in a subsequent health check:

```
ERROR   70.6s | RecoveryIT.testRecoveryWithConcurrentIndexing <<< FAILURES!
   > Throwable #1: org.elasticsearch.client.ResponseException: method [GET], host [http://[::1]:44918], URI [/_cluster/health/recovery_with_concurrent_indexing?wait_for_no_relocating_shards=true&level=shards&timeout=70s&wait_for_status=green], status line [HTTP/1.1 408 Request Timeout]
   > {"cluster_name":"rolling-upgrade","status":"yellow","timed_out":true,"number_of_nodes":3,"number_of_data_nodes":3,"active_primary_shards":1,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":98.68421052631578,"indices":{"recovery_with_concurrent_indexing":{"status":"yellow","number_of_shards":1,"number_of_replicas":2,"active_primary_shards":1,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"shards":{"0":{"status":"yellow","primary_active":true,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1}}}}}
   > 	at __randomizedtesting.SeedInfo.seed([DA5FC8F3EE50997:8DF438DCBC41D75D]:0)
   > 	at org.elasticsearch.client.RestClient$SyncResponseListener.get(RestClient.java:926)
   > 	at org.elasticsearch.client.RestClient.performRequest(RestClient.java:229)
   > 	at org.elasticsearch.test.rest.ESRestTestCase.ensureGreen(ESRestTestCase.java:711)
   > 	at org.elasticsearch.upgrades.RecoveryIT.testRecoveryWithConcurrentIndexing(RecoveryIT.java:141)
   > 	at java.lang.Thread.run(Thread.java:748)
   > Caused by: org.elasticsearch.client.ResponseException: method [GET], host [http://[::1]:44918], URI [/_cluster/health/recovery_with_concurrent_indexing?wait_for_no_relocating_shards=true&level=shards&timeout=70s&wait_for_status=green], status line [HTTP/1.1 408 Request Timeout]
   > {"cluster_name":"rolling-upgrade","status":"yellow","timed_out":true,"number_of_nodes":3,"number_of_data_nodes":3,"active_primary_shards":1,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":98.68421052631578,"indices":{"recovery_with_concurrent_indexing":{"status":"yellow","number_of_shards":1,"number_of_replicas":2,"active_primary_shards":1,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"shards":{"0":{"status":"yellow","primary_active":true,"active_shards":2,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1}}}}}
   > 	at org.elasticsearch.client.RestClient$1.completed(RestClient.java:546)
   > 	at org.elasticsearch.client.RestClient$1.completed(RestClient.java:531)
   > 	at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:119)
   > 	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:177)
   > 	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:436)
   > 	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:326)
   > 	at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
   > 	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
   > 	at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
   > 	at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:114)
   > 	at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
   > 	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
   > 	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
   > 	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
   > 	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
   > 	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
   > 	... 1 more
```
(from https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+intake/186/console)

Unfortunately the first few failures in the sequence of broken builds on 6.x seem unrelated to this one, with more changes having been pushed on top of the red CI, so it's hard to narrow down the point at which things started to break. Looking at the changes, I wonder if perhaps #34902 or #35332 could have this sort of effect?

Anyhow, muted in 188de54bed63e728dbd83419a8a5c209aa1c7f78.

--- a/server/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java
+++ b/server/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java
@@ -105,13 +105,13 @@ public class TransportResyncReplicationAction extends TransportWriteAction<Resyn
 
     @Override
     protected ClusterBlockLevel globalBlockLevel() {
-        // resync should never be blocked
+        // resync should never be blocked because it's an internal action
         return null;
     }
 
     @Override
     protected ClusterBlockLevel indexBlockLevel() {
-        // resync should never be blocked
+        // resync should never be blocked because it's an internal action
         return null;
     }
 

--- a/server/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java
+++ b/server/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java
@@ -27,6 +27,7 @@ import org.elasticsearch.action.support.replication.ReplicationResponse;
 import org.elasticsearch.action.support.replication.TransportReplicationAction;
 import org.elasticsearch.action.support.replication.TransportWriteAction;
 import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
 import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
 import org.elasticsearch.cluster.node.DiscoveryNode;
 import org.elasticsearch.cluster.routing.ShardRouting;
@@ -100,6 +101,18 @@ public class TransportResyncReplicationAction extends TransportWriteAction<Resyn
             final long pre60NodeCheckpoint = SequenceNumbers.PRE_60_NODE_CHECKPOINT;
             listener.onResponse(new ReplicaResponse(pre60NodeCheckpoint, pre60NodeCheckpoint));
         }
+    }
+
+    @Override
+    protected ClusterBlockLevel globalBlockLevel() {
+        // resync should never be blocked
+        return null;
+    }
+
+    @Override
+    protected ClusterBlockLevel indexBlockLevel() {
+        // resync should never be blocked
+        return null;
     }
 
     @Override

--- a/server/src/test/java/org/elasticsearch/action/resync/TransportResyncReplicationActionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/resync/TransportResyncReplicationActionTests.java
@@ -1,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.action.resync;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.block.ClusterBlocks;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.network.NetworkService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.discovery.DiscoverySettings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ReplicationGroup;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.MockTcpTransport;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+import java.nio.charset.Charset;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Collections.emptyList;
+import static org.elasticsearch.action.support.replication.ClusterStateCreationUtils.state;
+import static org.elasticsearch.test.ClusterServiceUtils.createClusterService;
+import static org.elasticsearch.test.ClusterServiceUtils.setState;
+import static org.elasticsearch.transport.TransportService.NOOP_TRANSPORT_INTERCEPTOR;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyObject;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TransportResyncReplicationActionTests extends ESTestCase {
+
+    private static ThreadPool threadPool;
+
+    @BeforeClass
+    public static void beforeClass() {
+        threadPool = new TestThreadPool("ShardReplicationTests");
+    }
+
+    @AfterClass
+    public static void afterClass() {
+        ThreadPool.terminate(threadPool, 30, TimeUnit.SECONDS);
+        threadPool = null;
+    }
+
+    public void testResyncDoesNotBlockOnPrimaryAction() throws Exception {
+        try (ClusterService clusterService = createClusterService(threadPool)) {
+            final String indexName = randomAlphaOfLength(5);
+            setState(clusterService, state(indexName, true, ShardRoutingState.STARTED));
+
+            setState(clusterService,
+                ClusterState.builder(clusterService.state()).blocks(ClusterBlocks.builder()
+                    .addGlobalBlock(DiscoverySettings.NO_MASTER_BLOCK_ALL)
+                    .addIndexBlock(indexName, IndexMetaData.INDEX_WRITE_BLOCK)));
+
+            try (MockTcpTransport transport = new MockTcpTransport(Settings.EMPTY, threadPool, BigArrays.NON_RECYCLING_INSTANCE,
+                new NoneCircuitBreakerService(), new NamedWriteableRegistry(emptyList()), new NetworkService(emptyList()))) {
+
+                final MockTransportService transportService = new MockTransportService(Settings.EMPTY, transport, threadPool,
+                    NOOP_TRANSPORT_INTERCEPTOR, x -> clusterService.localNode(), null, Collections.emptySet());
+                transportService.start();
+                transportService.acceptIncomingRequests();
+                final ShardStateAction shardStateAction = new ShardStateAction(clusterService, transportService, null, null, threadPool);
+
+                final IndexMetaData indexMetaData = clusterService.state().metaData().index(indexName);
+                final Index index = indexMetaData.getIndex();
+                final ShardId shardId = new ShardId(index, 0);
+                final IndexShardRoutingTable shardRoutingTable = clusterService.state().routingTable().shardRoutingTable(shardId);
+                final ShardRouting primaryShardRouting = clusterService.state().routingTable().shardRoutingTable(shardId).primaryShard();
+                final String allocationId = primaryShardRouting.allocationId().getId();
+                final long primaryTerm = indexMetaData.primaryTerm(shardId.id());
+
+                final IndexShard indexShard = mock(IndexShard.class);
+                when(indexShard.shardId()).thenReturn(shardId);
+                when(indexShard.routingEntry()).thenReturn(primaryShardRouting);
+                when(indexShard.getPendingPrimaryTerm()).thenReturn(primaryTerm);
+                doAnswer(invocation -> {
+                    ActionListener<Releasable> callback = (ActionListener<Releasable>) invocation.getArguments()[0];
+                    callback.onResponse(() -> logger.trace("released"));
+                    return null;
+                }).when(indexShard).acquirePrimaryOperationPermit(any(ActionListener.class), anyString(), anyObject());
+                when(indexShard.getReplicationGroup()).thenReturn(
+                    new ReplicationGroup(shardRoutingTable,
+                        clusterService.state().metaData().index(index).inSyncAllocationIds(shardId.id()),
+                        shardRoutingTable.getAllAllocationIds()));
+
+                final IndexService indexService = mock(IndexService.class);
+                when(indexService.getShard(eq(shardId.id()))).thenReturn(indexShard);
+
+                final IndicesService indexServices = mock(IndicesService.class);
+                when(indexServices.indexServiceSafe(eq(index))).thenReturn(indexService);
+
+                final IndexNameExpressionResolver resolver = new IndexNameExpressionResolver();
+                final TransportResyncReplicationAction action = new TransportResyncReplicationAction(Settings.EMPTY, transportService,
+                    clusterService, indexServices, threadPool, shardStateAction, new ActionFilters(new HashSet<>()), resolver);
+
+                assertThat(action.globalBlockLevel(), nullValue());
+                assertThat(action.indexBlockLevel(), nullValue());
+
+                final Task task = mock(Task.class);
+                when(task.getId()).thenReturn(randomNonNegativeLong());
+
+                final byte[] bytes = "{}".getBytes(Charset.forName("UTF-8"));
+                final ResyncReplicationRequest request = new ResyncReplicationRequest(shardId, 42L, 100,
+                    new Translog.Operation[]{new Translog.Index("type", "id", 0, primaryTerm, 0L, bytes, null, -1)});
+
+                final PlainActionFuture<ResyncReplicationResponse> listener = new PlainActionFuture<>();
+                action.sync(request, task, allocationId, primaryTerm, listener);
+
+                assertThat(listener.get().getShardInfo().getFailed(), equalTo(0));
+                assertThat(listener.isDone(), is(true));
+            }
+        }
+    }
+}

